nucliadb_node/nucliadb_node/shadow_shards.py

```diff
@@ -0,0 +1,253 @@
+        if not self._folder.endswith("/"):
+            self._folder += "/"
+        await aos.makedirs(self._folder, exist_ok=True)
+
+        await self.load_metadata()
+        await self.load_shards()
+        self._loaded = True
+
+    async def load_shards(self):
+        self.shards = set()
+        for shard_path in glob.glob(f"{self._folder}/*"):
+            shard_id = shard_path.split(self._folder)[-1].lstrip("/")
+            if shard_id != self._metadata_file:
+                self.shards.add(shard_id)
+
+    async def load_metadata(self) -> None:
+        metadata_path = self.shard_path(self._metadata_file)
+        self._metadata = await ShadowMetadata.load(metadata_path)
+
+    async def save_metadata(self) -> None:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        metadata_path = self.shard_path(self._metadata_file)
+        await self.metadata.save(metadata_path)
+
+    async def increment_ops(self, shard_id: str) -> None:
+        self.metadata.increment_ops(shard_id)
+        await self.save_metadata()
+
+    @property
+    def loaded(self) -> bool:
+        return self._loaded
+
+    @property
+    def metadata(self) -> ShadowMetadata:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        return self._metadata
+
+    def shard_path(self, shard_id: str) -> str:
+        return f"{self._folder}/{shard_id}"
+
+    async def create(self) -> str:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+
+        # Get a unique shard id
+        shard_id = uuid.uuid4().hex
+        while shard_id in self.shards:
+            shard_id = uuid.uuid4().hex
+
+        async with aiofiles.open(self.shard_path(shard_id), mode="x"):
+            pass
+
+        self.shards.add(shard_id)
+        self.metadata.shards[shard_id] = ShadowShardInfo(shard_id=shard_id)
+        await self.save_metadata()
+        return shard_id
+
+    async def delete(self, shard_id: str) -> None:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+
+        if shard_id not in self.shards:
+            raise ShadowShardNotFound()
+
+        shard_path = self.shard_path(shard_id)
+        await aiofiles.os.remove(shard_path)
+        self.shards.remove(shard_id)
+
+        self.metadata.shards.pop(shard_id)
+        await self.save_metadata()
+
+    def exists(self, shard_id: str) -> bool:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        return shard_id in self.shards
+
+    def encode_operation(self, node_operation: NodeOperation) -> bytes:
+        opcode, payload = node_operation
+        if opcode == OperationCode.SET:
+            payload = payload.SerializeToString()  # type: ignore
+        elif opcode == OperationCode.DELETE:
+            payload = payload.encode()  # type: ignore
+        else:
+            ValueError(f"Unknown opcode: {opcode}")
+        encoded = base64.b64encode(payload).decode()  # type: ignore
```

Comment:
Why you encode in base64 ? We could just store the PBs on disc

Reply:
Due to this: https://stackoverflow.com/questions/54397600/new-line-character-in-serialized-messages

Some brain resources, when serialised to string, will include newline characters, which causes trouble for us when reading and writing to the file.